adr/ADR-52.md

```diff
@@ -0,0 +1,124 @@
+  replicate asynchronously, so letting them answer would make the response take even longer to return.)
+- Mirrors can still answer `DirectGet` requests, the transparency of mirrors answering read requests will violate any
+  read-after-write guarantees (as the client will not know). This would mean mirrors must not be enabled if this
+  guarantee should be kept.
+- Read-after-write guarantees could temporarily be violated when scaling streams up or down.
+- This is not a compatible approach for consumers, meaning they could not have these guarantees based on this approach.
+
+Although having reads be served through Raft does (mostly) offer a strong guarantee of read-after-write and monotonic
+reads, the disadvantages outway the advantages. Ideally, the solution has the following advantages:
+
+- It's explicitly defined, either in configuration or in code.
+- Works for both replicated and non-replicated streams. (Scale up/down has no influence, and implementation is not
+  replication-specific)
+- Incurs no slowdown, just as fast as reads that don't guarantee read-after-write (no prior replication required).
+- Let followers, and even mirrors, answer read requests as long as they can make the guarantee.
+- Let followers, and mirrors, inform the client when they can't make the guarantee. The guarantee is always kept, but
+  an error is returned that can be retried (to get a successful read). This can be tuned by disabling reads on mirrors
+  or followers.
+
+Now, on to the proposed design which has the above advantages.
+
+The write and read paths remain eventually consistent as it is now. But one can opt-in for immediate consistency to
+guarantee read-after-write and monotonic reads, for both direct/msg read requests as well as consumers.
+
+- **Read-after-write** is achieved because all writes through `js.Publish`, `kv.Put`, etc. return the sequence
+  (inherently last sequence) of the stream. In `DirectGet` requests those observed last sequences can be used for read
+  requests.
+- **Monotonic reads** is achieved by collecting the highest sequence seen in read requests and using that sequence for
+  subsequent read requests.
+
+This can be implemented with an additional `MinLastSeq` field in `JSApiMsgGetRequest` and `ConsumerConfig`.
+
+- This ensures the server only replies with data if it can actually 100% guarantee immediate consistency. This is done
+  by confirming the `LastSeq` it has for its local stream, is at least the `MinLastSeq` specified.
+- Side-note: although `MsgGet` is only answered by the leader, technically an old leader could still respond and serve
+  stale reads. Although this shouldn't happen often in practice, until now we couldn't guarantee it. The error can be
+  detected on the old leader, and it can delay the error response, allowing for the real leader to send the actual
+  answer.
+- Followers/mirrors reject the read request if they can't satisfy the `MinLastSeq`. But can serve reads and share the
```

Comment:
I think overall this is quite elegant and solves the problem well, I have a few concerns though:

 * I think as stated here, in the face of lag or mirror direct, this will cause a lot of read errors, users will try to implement their own retry strategies and we will end up with a retry flood as naive implementations do so without backoff.  
 * I liked the earlier approach of forwarding the denied request to the leader, but this of course wont work for mirror direct as that will address the above issue elegantly and minimal client impact
 * I dont think mirror direct is broadly used but if we feel it is we should maybe consider a built-in retry strategy in clients when r-a-w is requested, this way we can enforce backoffs etc

Reply:
That might make assumptions about topology that we cant really make - mirrors cross account etc, though I imagine mirror direct as a option on those would be pretty weird

Reply:
I think so, have to figure out if we need to do something about naive retrying clients as per my points above for those

Reply:
Yeah thats an interesting idea

Reply:
Updated the ADR to:
- delay error responses, similar to how we delay JS API error responses
- redirect read requests to the leader, IFF a follower on the original (non-mirrored) stream